### PR TITLE
TMDM-10659	 Continued select count(*) queries are slow on large volumes

### PR DESCRIPTION
**Pull request by lpw82gf**
1, Add count data to cache to avoid frequently query, data saved in map, with Key={storageName}#{storageType}.{entityName}/{condition hash}, like `Product#MASTER.Product/3214568`
2, When data deleted/updated/created, after transaction committed, reset cache
3, When data cluster dropped, reset all the count cache belongs to this cluster
4, Standalone server using In Memory approach, while cluster environment using Hazelcast approach, like AutoIncrement.

### CODE REVIEW

File: org.talend.mdm.core.storage.sql/src/test/java/com/amalto/core/query/LongTransactionAbstractTestCase.java [hunk 1]
```diff
@@ -38,7 +39,7 @@
     
     @Override
     public void setUp() throws Exception {
-        // override to avoid storage.begin in superclass setup
+        EntityCountUtil.clearAll();
```
Comment (pbabin) [inline org.talend.mdm.core.storage.sql/src/test/java/com/amalto/core/query/LongTransactionAbstractTestCase.java]: Can the previous comment be kept. It is important one to understand why we don't call super.setUp

File: org.talend.mdm.core/src/com/amalto/core/storage/InMemoryCounter.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+package com.amalto.core.storage;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.amalto.core.storage.Counter.AbstractCounter;
+
+public class InMemoryCounter extends AbstractCounter implements Counter {
```
Comment (pbabin) [inline org.talend.mdm.core/src/com/amalto/core/storage/InMemoryCounter.java]: Both implementations of AbstractCounter are using an internal Map<String, Integer> . Cant' it be then part of the base class, thus simplifying a lot their implementations?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorage.java [hunk 3]
```diff
@@ -141,6 +141,9 @@
 import com.amalto.core.server.MetadataRepositoryAdmin;
 import com.amalto.core.server.ServerContext;
 import com.amalto.core.server.StorageAdmin;
+import com.amalto.core.storage.Counter.CountKey;
+import com.amalto.core.storage.Counter.CountKeyList;
```
Comment (htyin) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorage.java]: CountKeyList is never used.

Comment (lpw82gf) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorage.java]: It is used in `EntityCountUtil#addToClearList`

**Pull request closed without merging**